### ISSUE TITLE
exception during rendition generation breaks image chooser and image admin views that attempt to display the image

### ISSUE THREAD

**Issue by thenewguy** [torchbox/wagtail#1792]
Wagtail will let you upload the following image:
![shut-up-and-take-my-money-animated](https://cloud.githubusercontent.com/assets/77731/10294719/0c7cfae0-6b8a-11e5-97fe-5ef772b24e38.gif)

However it cannot be used to create a rendition.  `image.get_rendition(...)` does not catch the exception raised by PIL GifImagePlugin "int() argument must be a string or a number, not 'tuple'".

This breaks the image chooser and anything that uses `get_rendition()`.

There are some older issues related to this that are closed and an open one on willow at https://github.com/torchbox/Willow/issues/16

Wagtail needs to handle a failed rendition more gracefully because currently the interface breaks and becomes unusable with 500 errors.  Possibly return a rendition that points to a static error image available from the site's static files indicating a problem.  I am going to investigate this approach with a custom image model and see how it goes.

**Comment by thenewguy** [comment]
It also might not be a bad idea for the upload validator to make sure a rendition can be created

**Comment by thenewguy** [comment]
So something like the following works around the issue:
```
class ExceptionImage(object):
    title="Unrecoverable error encountered producing this image"

class ExceptionRendition(AbstractRendition):
    class Meta:
        abstract = True
    
    def __init__(self, *args, **kwargs):
        super(ExceptionRendition, self).__init__(*args, **kwargs)
        self.image, self.width, self.height = ExceptionImage(), 80, 80
    
    @property
    def url(self):
        return staticfiles_storage.url("images/rendition-exception.png")

class TheImageModel(AbstractImage):

	...

	def get_rendition(self, *args, **kwargs):
		try:
			rendition = super(TheImageModel, self).get_rendition(*args, **kwargs)
		except Exception:
			rendition = ExceptionRendition()
		return rendition

**Comment by thenewguy** [comment]
Also I realize animation is the problem.  I've seen the notice at http://docs.wagtail.io/en/v1.1/topics/images/advanced_topics.html#animated-gif-support to use the Wand library.  But I think the rendition generation should be a little more robust so that the image chooser and image admin do not fail with 500 errors if animated gif is uploaded when wand isn't installed.

**Comment by benje** [comment]
Also encountering this issue, any help/fix would be appreciated as the image admin is now breaking :(

**Comment by gasman** [comment]
@benje - https://groups.google.com/d/msg/wagtail/CQ-S338XDvs/sYm11mjJCQAJ might help you out.

**Comment by benje** [comment]
Thanks @gasman

**Comment by chhantyal** [comment]
Thanks @thenewguy 

The docs you referenced are right. I had to install `Wand` for GIF and it is fixed.

**Comment by ghostwords** [comment]
I ran into the same issue as @thenewguy, with Wand installed. Strangely, GIFs worked on my local machine, but not in production. Python versions and packages matched ... Looks like the difference was the `imagemagick` package, installed locally but not remotely.

```
$ python -m wand.version --verbose
Wand 0.4.1
ImageMagick 6.7.7-10 2014-03-06 Q16 http://www.imagemagick.org
```

Installing `imagemagick` seems to have fixed GIF uploading. Processing new GIFs can take a long time, however. For example, [this one](http://media2.giphy.com/media/SuJoKxDplnKve/giphy.gif) took about seven (!) minutes from initial upload to being able to open its image details page successfully:

```
2015-11-05 15:58:22.381485545 -0500 ./images/lemon.max-800x600.gif
2015-11-05 15:55:30.513605457 -0500 ./images/lemon.original.gif
2015-11-05 15:53:20.549696444 -0500 ./images/lemon.max-165x165.gif
2015-11-05 15:51:09.305788613 -0500 ./original_images/lemon.gif
```

While the image is being processed, there is no indication of progress, just a page that won't load. Looks like there are two delays: one requesting the top-level images index page (smaller delay), then another requesting the image details page (larger delay).

Is this a performance bug? Is it also a documentation bug (how should one know to install `imagemagick`)? As mentioned above, there is also the issue that failure to process an image results in a broken CMS.

**Comment by shredding** [comment]
This may be becauce of this bug i reported:

https://github.com/torchbox/Willow/pull/23

**Comment by shredding** [comment]
... And fixed in the PR

**Comment by shredding** [comment]
@gasman: https://github.com/torchbox/Willow/pull/23 has been merged. Will it be possible to have a version including that fix be part of 1.2?

**Comment by gasman** [comment]
@shredding 1.2 was finalised last night, I'm afraid. Is there anything that needs to be done on Wagtail's side to fix this, other than requiring a newer version of Willow (when it's released)?

(@kaedroho - any objection to us specifying a wider version range in https://github.com/torchbox/wagtail/blob/23ed5392de1dfc67031aa04190c2097d74f7ae5c/setup.py#L34 ? We can do something like `Willow>=0.2.1,<1.0` if you think there might be backwards-incompatible changes on the way...)

**Comment by shredding** [comment]
@gasman : Great news. There is nothing to do on wagtails side besides using latest willow.

**Comment by gasman** [comment]
@shredding Thanks for confirming! Have marked this as a 1.3 bug now, so that we remember to put out a new Willow release in time for Wagtail 1.3...

**Comment by shredding** [comment]
Okay, so for the googlers:

Patch your pillow.py in willow like this until the error is fixed in 1.3:
https://github.com/torchbox/Willow/blob/master/willow/backends/pillow.py#L83-L84

**Comment by kaedroho** [comment]
@gasman Yep, we can do that. I'd like to have the freedom to refine the API before the 1.0 release so could we pin it to ``<0.3``?

**Comment by gasman** [comment]
@kaedroho Sure - sounds like a plan!

**Comment by kaedroho** [comment]
I've just released Willow 0.2.2 on pypi. If you add this to your requirements.txt, it should override the version specified in Wagtail's setup.py

**Comment by gasman** [comment]
Willow version-bumped in 6e4cd12e24a69a2a497be8d81795e570ad81c6a3; have confirmed that it now uploads the image above successfully using the Pillow backend (albeit without the animation).
